<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.37">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Devan Becker">
<meta name="dcterms.date" content="2025-05-07">

<title>Simulating Wastewater SARS-CoV-2 Reads with SWAMPy – Devan Nightly</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-29e2c20b02301cfff04dc8050bf30c7e.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-a5a7a9aa3d5b07c7c3601f2586d7e9ee.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Devan Nightly</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">About</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Simulating Wastewater SARS-CoV-2 Reads with SWAMPy</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">code</div>
                <div class="quarto-category">bioinformatics</div>
              </div>
                  </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Devan Becker </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">May 7, 2025</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<p>Flushing a collection of genomic sequences down the toilet with <a href="https://github.com/goldman-gp-ebi/SWAMPy">SWAMPy</a>.</p>
<p>The overall goal of SWAMPy is to simulate wastewater short read data (in fastq format) based on known abundances of genomic sequences. There are a couple important points here:</p>
<ul>
<li>Sequences, not lineages. In general, there isn’t a “canonical” sequence for a lineage. Lineages are defined according to placement on a phylogenetic tree.</li>
<li>There are a lot of assumptions required to simulate these data. The amount of degredation, random sequencing errors, and the viral load are all random phenomena.</li>
</ul>
<section id="test-run" class="level2">
<h2 class="anchored" data-anchor-id="test-run">Test Run</h2>
<p>The following code and files come from the SWAMPy repo, this is just a verification that I can run it on my machine. In order for the output to appear, I’m running shell commands from within python.</p>
<p>First, let’s take a look at the example files provided in the <code>example</code> folder of the SWAMPy repo. The <code>genomes.fasta</code> file has the full genome sequences in it, so let’s just display the lines that start wiht “<code>&gt;</code>”, i.e.&nbsp;the names of the sequences.</p>
<div id="57fe5ddc" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>os.getcwd()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>genome_names <span class="op">=</span> subprocess.run(</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    [<span class="st">"grep"</span>, <span class="st">"&gt;"</span>, <span class="st">"../../data/genomes.fasta"</span>],</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    capture_output <span class="op">=</span> <span class="va">True</span>, text <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(genome_names.stdout)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&gt;OA984024.1 Severe acute respiratory syndrome coronavirus 2 genome assembly, complete genome: monopartite
&gt;OD929257.1 Severe acute respiratory syndrome coronavirus 2 genome assembly, complete genome: monopartite
&gt;OM389144.1 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/GA-EHC-111G/2020, complete genome
</code></pre>
</div>
</div>
<p>In the <code>abundances.tsv</code> file, notice that these names match exactly (up to differences in formatting). The “50”, “40”, and “10” at the end of the line are a different column (because this is a <code>.tsv</code> file, columns are separated by tab characters). Note that the code below shows the entire contents of the file; this is all that is needed.</p>
<div id="ece6d05e" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(<span class="st">"../../data/abundances.tsv"</span>) <span class="im">as</span> f:</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> line <span class="kw">in</span> f:</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>      <span class="bu">print</span>(line.strip())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>OA984024.1 Severe acute respiratory syndrome coronavirus 2 genome assembly, complete genome: monopartite    50
OD929257.1 Severe acute respiratory syndrome coronavirus 2 genome assembly, complete genome: monopartite    40
OM389144.1 Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/GA-EHC-111G/2020, complete genome   10</code></pre>
</div>
</div>
</section>
<section id="running-swampy" class="level2">
<h2 class="anchored" data-anchor-id="running-swampy">Running SWAMPy</h2>
<p>The following code will actually run the SWAMPy simulation routine. In the code below, there are many options set. Most of them are self-explanatory, but I’ll go into some detail for my own sake.</p>
<ul>
<li>A primer set is what is used to grab the sequences in the wastewater. The the ARTIC protocol, these are about 400 base pair portions of the genome, where the ends of the primers grab the sequence and then the middles are read by the sequencer.
<ul>
<li>The ARTIC protocol has several <a href="https://github.com/artic-network/primer-schemes">primer schemes</a>. The <code>--primer_set</code> cli argument to swampy allows the user to decide on the primer set, with options defined in the <a href="https://github.com/goldman-gp-ebi/SWAMPy/wiki/CLI-arguments">swapy wiki</a>.</li>
<li>The primer set probably won’t affect the results too much as long as the correct primer set is provided when processing the simulated reads.</li>
</ul></li>
<li>The <code>--n_reads</code> parameter works as it should, but it’s not necessarily intuitive.
<ul>
<li>One person can poop hundreds or thousands of whole or partial genomes.</li>
<li>These genomes degrade in the wastewater, so not all portions can be read.</li>
<li>Each of those genomes could have up to 196 reads<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> in the ARTIC protocol.</li>
<li>Some primers will fail to be read (amplicon dropout).</li>
<li>Because of this, the number of reads is a target number, and the actual simulation may result in a slightly different number.</li>
</ul></li>
</ul>
<div id="5c7eae11" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> subprocess</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>swampy_executable <span class="op">=</span> os.path.expanduser(<span class="st">"~/git/DASL/external/SWAMPy/src/simulate_metagenome.py"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>output <span class="op">=</span> subprocess.run(</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"python"</span>, swampy_executable,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--primer_set"</span>, <span class="st">"a1"</span>,</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--genomes_file"</span>, <span class="st">"../../data/genomes.fasta"</span>,</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--genome_abundances"</span>, <span class="st">"../../data/abundances.tsv"</span>,</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--n_reads"</span>, <span class="st">"1000"</span>,</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--output_folder"</span>, <span class="st">"simulation_output"</span>,</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--temp_folder"</span>, <span class="st">"temp"</span>,</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--autoremove"</span>,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">"--quiet"</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    ]</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(output.stderr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>2025-05-07 10:07:11,059 [INFO] Primer set: Artic v1
2025-05-07 10:07:11,059 [INFO] Number of reads: 1000
2025-05-07 10:07:11,060 [INFO] Random seed: 809554866
2025-05-07 10:07:11,060 [INFO] Amplicon pseudocounts/ i.e. quality parameter: 200
2025-05-07 10:07:11,172 [INFO] Total of relative abundance values is 100.0, not 1.
2025-05-07 10:07:11,172 [INFO] Continuing, normalising total of genome abundances to 1.
Building a SMALL index
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.3.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.3.bt2
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.4.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.4.bt2
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.1.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.1.bt2
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.2.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.2.bt2
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.1.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.1.bt2
Renaming temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.2.bt2.tmp to temp/indices/OA984024.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.2.bt2
Building a SMALL index
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.3.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.3.bt2
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.4.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.4.bt2
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.1.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.1.bt2
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.2.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.2.bt2
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.1.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.1.bt2
Renaming temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.2.bt2.tmp to temp/indices/OD929257.1_Severe_acute_respiratory_syndrome_coronavirus_2_genome_assembly&amp;_complete_genome:_monopartite.rev.2.bt2
Building a SMALL index
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.3.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.3.bt2
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.4.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.4.bt2
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.1.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.1.bt2
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.2.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.2.bt2
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.rev.1.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.rev.1.bt2
Renaming temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.rev.2.bt2.tmp to temp/indices/OM389144.1_Severe_acute_respiratory_syndrome_coronavirus_2_isolate_SARS-CoV-2&amp;human&amp;USA&amp;GA-EHC-111G&amp;2020&amp;_complete_genome.rev.2.bt2
2025-05-07 10:07:52,898 [INFO] Generating reads using art_illumina, cycling through all genomes and remaining amplicons.
2025-05-07 10:07:53,545 [INFO] Creating random data for shuffle.
2025-05-07 10:07:54,964 [INFO] Shuffling simulation_output/example_R1.fastq
2025-05-07 10:07:55,105 [INFO] Shuffling simulation_output/example_R2.fastq</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>None</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2025-05-07 10:07:55,232 [INFO] Exiting sars-cov-2 metagenome simulator - tidying up.
2025-05-07 10:07:55,255 [INFO] Removing all files in temp/genomes
2025-05-07 10:07:55,257 [INFO] Removing all files in temp/amplicons
2025-05-07 10:07:55,317 [INFO] Removing all files in temp/indices</code></pre>
</div>
</div>
</section>
<section id="the-simulation-model" class="level2">
<h2 class="anchored" data-anchor-id="the-simulation-model">The Simulation Model</h2>
<p>The default model for the number of observations of each amplicon is labelled <code>DIRICHLET_1</code>. Based on digging through the code, it works as follows:</p>
<ol type="1">
<li>Choose the number of each supplied sequence to use (multinomial).
<ul>
<li>Given the total reads (<code>--n_reads</code>) and the abundances, it generates the number of each supplied genome according to a multinomial distribution.</li>
</ul></li>
<li>Choose the probability of each amplicon (dirichlet * psuedocounts).
<ul>
<li>The probability of choosing each amplicon is defined by a Dirichlet distribution. This uses a set of hyperparameters to define the <span class="math inline">\(\underline\alpha\)</span> parameter, then samples a new set of probabilities.</li>
<li>Depending on the primer set, different hyperparameters are assumed. For built-in primer sets, there are files defining these hyperparameters.</li>
<li>The pseudocount essentially controls the variance of the dirichlet distribution. A higher pseudocount makes the simulated probabilities closer to the hyperparameters.</li>
</ul></li>
<li>Sample the amplicons from each genome sequence (multinomial for each genome, based on output of dirichlet)
<ul>
<li>The code essentially goes through a for loop (dictionary comprehension), and for each supplied genome it samples amplicons.</li>
</ul></li>
</ol>
<p>The model labelled <code>DIRICHLET_2</code> appears to be the same as <code>DIRICHLET_1</code> but does not use pseudocounts. The <code>EXACT</code> model just uses the abundances provided.</p>
<p>This process would result in a collection of amplicons that all look exactly the same. This is not realistic - high-throughput sequence tends to produce many errors, but makes up for it by getting a lot of reads.</p>
<p>To get these errors, there are several parameters that define genome-wide error rates for insertions, deletions, and substitutions, as well as dirichlet hyper-parameters to ensure that the probabilies provided also have some variance. Alternative alleles are generated for each amplicon, then the number of each alternative allele is chosen based on a multinomial distribution (which, in this case, is a binomial since there’s only the mutation and the original). See the <a href="https://github.com/goldman-gp-ebi/SWAMPy/wiki/CLI-arguments">CLI arguments wiki</a> and the <a href="https://github.com/goldman-gp-ebi/SWAMPy/wiki/SWAMPy-method#4-high-frequency-error-introduction">methods</a>.</p>
</section>
<section id="conclusion" class="level2">
<h2 class="anchored" data-anchor-id="conclusion">Conclusion</h2>
<p>A great deal of work has gone into ensureing that short are simulated in a reasonable manner and then correctly placed in a fastq file. I commend the authors on this codebase!</p>
<p>The simulation process appears to be reasonable, but choice of the relevant parameters might be tricky. Choice of the parameters will greatly affect the variance of the output, which is the main goal of my research. I plan to use the default values for now, but I will be careful to speak in relative terms. I’m certainly not getting the absolute variance of the estimators - there will always be aleatoric uncertainty, despite our best efforts. However, if I simulate a lot of data sets and get an estimator that tends to have a lower variance, I can be confident that it is lower (but I can’t say that the estimated value is the true value).</p>
</section>
<section id="other-lessons" class="level2">
<h2 class="anchored" data-anchor-id="other-lessons">Other Lessons</h2>
<p>While making this post, I had so many “opportunities for learning” (i.e., frustrations).</p>
<ol type="1">
<li>To set up a python environment that Quarto will see, I need to use <code>conda create --prefix ./dn_env python=3.10</code> to ensure that it’s in the current directory.
<ul>
<li>I need to force VSCode to activate it by selecting the interpreter.</li>
<li>After creating the env, I need to activate it and run <code>conda install nbconvert jupyter jupyter-cache biopython pandas</code></li>
<li>I <strong>do not</strong> need to activate the environment from within the quarto document.</li>
</ul></li>
<li><code>subprocess.run()</code> needs a list, with each argument being a different element (no spaces).</li>
</ol>


</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>There are 98 primers in the ARTIC protocol, and each read can be read forward and backwards.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>